File: packages/xcon/src/parser.ts
import {
  XCONDocument,
  HeaderNode,
  BodyNode,
  RowNode,
  DocumentNode,
  FieldValueNode,
  LabelNode,
  ValueNode,
  ArrayNode,
  ParseOptions,
} from './ast';
import { tokenize, Token, TokenType } from './tokenizer';
import { XCONParseError } from './errors';
import { inferType } from './evaluator';

class Parser {
  private tokens: Token[];

  private pos: number = 0;

  constructor(tokens: Token[]) {
    this.tokens = tokens;
  }

  private current(): Token {
    return this.tokens[this.pos]!;
  }

  private peek(offset = 1): Token | undefined {
    return this.tokens[this.pos + offset];
  }

  private advance(): Token {
    return this.tokens[this.pos++]!;
  }

  private expect(type: TokenType): Token {
    const token = this.current();
    if (token.type !== type) {
      throw new XCONParseError(
        `Expected ${type} but got ${token.type}`,
        token.line,
        token.column,
      );
    }
    return this.advance();
  }

  private skipNewlines(): void {
    while (this.current().type === TokenType.NEWLINE) {
      this.advance();
    }
  }

  parse(): XCONDocument {
    this.skipNewlines();

    let header: HeaderNode | null = null;

    // Check for header
    if (this.current().type === TokenType.LPAREN) {
      header = this.parseHeader();
      this.skipNewlines();
    }

    const body = this.parseBody();

    return {
      type: 'XCONDocument',
      header,
      body,
    };
  }

  private parseHeader(): HeaderNode {
    this.expect(TokenType.LPAREN);
    const labels = this.parseLabelList();
    this.expect(TokenType.RPAREN);
    return { type: 'Header', labels };
  }

  private parseLabelList(): LabelNode[] {
    const labels: LabelNode[] = [];

    labels.push(this.parseLabel());

    while (this.current().type === TokenType.COMMA) {
      this.advance();
      labels.push(this.parseLabel());
    }

    return labels;
  }

  private parseLabel(): LabelNode {
    const token = this.expect(TokenType.LABEL_OR_VALUE);
    let isArray = false;
    let children: LabelNode[] = [];

    if (this.current().type === TokenType.ARRAY_MARKER) {
      this.advance();
      isArray = true;
    }

    // Check for nested schema: label:(sublabel,...)
    if (this.current().type === TokenType.COLON) {
      this.advance();
      if (this.current().type === TokenType.LPAREN) {
        this.advance();
        children = this.parseLabelList();
        this.expect(TokenType.RPAREN);
      }
    }

    return {
      type: 'Label',
      name: token.value,
      isArray,
      children,
    };
  }

  private parseBody(): BodyNode {
    const rows: RowNode[] = [];

    while (this.current().type !== TokenType.EOF) {
      this.skipNewlines();
      if (this.current().type === TokenType.EOF) break;

      rows.push(this.parseRow());
      this.skipNewlines();
    }

    return { type: 'Body', rows };
  }

  private parseRow(): RowNode {
    let label: string | null = null;

    // Check for row label (LABEL_OR_VALUE followed by COLON)
    if (
      this.current().type === TokenType.LABEL_OR_VALUE &&
      this.peek()?.type === TokenType.COLON
    ) {
      label = this.advance().value;
      this.expect(TokenType.COLON);
    }

    // Allow bare array as row body, not just a document
    let document: DocumentNode;
    if (this.current().type === TokenType.LBRACKET) {
      const array = this.parseArray();
      document = { type: 'Document', fields: [array] };
    } else {
      document = this.parseDocument();
    }

    return {
      type: 'Row',
      label,
      document,
    };
  }

  private parseDocument(): DocumentNode {
    this.expect(TokenType.LBRACE);

    const fields: FieldValueNode[] = [];

    while (this.current().type !== TokenType.RBRACE) {
      fields.push(this.parseFieldValue());

      if (this.current().type === TokenType.COMMA) {
        this.advance();
        // Allow trailing comma before close brace is an error
        if (this.current().type === TokenType.RBRACE) {
          throw new XCONParseError(
            'Trailing comma not allowed',
            this.current().line,
            this.current().column,
          );
        }
      }
    }

    this.expect(TokenType.RBRACE);

    return {
      type: 'Document',
      fields,
    };
  }

  private parseFieldValue(): FieldValueNode {
    if (this.current().type === TokenType.LBRACE) {
      return this.parseDocument();
    }

    if (this.current().type === TokenType.LBRACKET) {
      return this.parseArray();
    }

    // Empty value case
    if (this.current().type === TokenType.COMMA || this.current().type === TokenType.RBRACE) {
      return {
        type: 'Value',
        raw: '',
      };
    }

    // Must be a value (bare word or quoted string)
    if (
      this.current().type === TokenType.LABEL_OR_VALUE ||
      this.current().type === TokenType.QUOTED_STRING
    ) {
      const token = this.advance();
      return {
        type: 'Value',
        raw: token.value,
      };
    }

    throw new XCONParseError(
      `Unexpected token: ${this.current().type}`,
      this.current().line,
      this.current().column,
    );
  }

  private parseArray(): ArrayNode {
    this.expect(TokenType.LBRACKET);

    const items: FieldValueNode[] = [];

    while (this.current().type !== TokenType.RBRACKET) {
      items.push(this.parseFieldValue());

      if (this.current().type === TokenType.COMMA) {
        this.advance();
        if (this.current().type === TokenType.RBRACKET) {
          throw new XCONParseError(
            'Trailing comma not allowed',
            this.current().line,
            this.current().column,
          );
        }
      }
    }

    this.expect(TokenType.RBRACKET);

    return {
      type: 'Array',
      items,
    };
  }
}

export function parseToAST(input: string, _options?: ParseOptions): XCONDocument {
  const tokens = tokenize(input);
  const parser = new Parser(tokens);
  return parser.parse();
}

export function parse(input: string, options?: ParseOptions): Record<string, unknown> | unknown[] {
  const ast = parseToAST(input, options);
  return evaluate(ast);
}

function evaluate(
  doc: XCONDocument,
): Record<string, unknown> | unknown[] {
  if (doc.body.rows.length === 0) {
    return doc.header ? {} : [];
  }

  // Check if rows have labels
  const hasLabels = doc.body.rows.some((r) => r.label !== null);
  const allHaveLabels = doc.body.rows.every((r) => r.label !== null);

  if (hasLabels && !allHaveLabels) {
    throw new XCONParseError('Mixed labeled and unlabeled rows', 0, 0);
  }

  if (hasLabels) {
    const result: Record<string, unknown> = {};
    for (const row of doc.body.rows) {
      const value = evaluateDocument(row.document, doc.header?.labels ?? []);
      result[row.label!] = value;
    }
    return result;
  }

  // No labels - return array
  const result: unknown[] = [];
  for (const row of doc.body.rows) {
    const value = evaluateDocument(row.document, doc.header?.labels ?? []);
    result.push(value);
  }
  return result;
}

function evaluateDocument(
  doc: DocumentNode,
  schema: LabelNode[],
): Record<string, unknown> | unknown[] {
  const values: unknown[] = [];

  for (const field of doc.fields) {
    values.push(evaluateFieldValue(field, schema[values.length]?.children ?? []));
  }

  if (schema.length === 0) {
    // No schema: return array
    return values;
  }

  // Apply schema to values positionally
  const result: Record<string, unknown> = {};
  for (let i = 0; i < schema.length; i++) {
    const label = schema[i];
    const value = values[i];

    if (!label) continue;

    if (label.isArray) {
      // This field should be an array
      if (Array.isArray(value)) {
        result[label.name] = value;
      } else if (value === undefined) {
        result[label.name] = [];
      } else {
        result[label.name] = [value];
      }
    } else if (label.children.length > 0) {
      // This field is a nested document
      result[label.name] = value;
    } else {
      result[label.name] = value;
    }
  }

  return result;
}

function evaluateFieldValue(
  field: FieldValueNode,
  schema: LabelNode[],
): unknown {
  if (field.type === 'Value') {
    return inferType(field.raw);
  }

  if (field.type === 'Array') {
    return field.items.map((item) => evaluateFieldValue(item, []));
  }

  if (field.type === 'Document') {
    return evaluateDocument(field, schema);
  }

  return null;
}

File: packages/lmon-python/lmon/parser.py
from typing import List, Dict, Union, Any, Optional
from .tokenizer import Token, TokenType, tokenize
from .ast_nodes import (
    LMONDocument,
    HeaderNode,
    BodyNode,
    RowNode,
    DocumentNode,
    FieldValueNode,
    LabelNode,
    ValueNode,
    ArrayNode,
)
from .errors import LMONParseError
from .evaluator import infer_type


class Parser:
    """Recursive descent parser for LMON."""

    def __init__(self, tokens: List[Token]) -> None:
        self.tokens = tokens
        self.pos = 0

    def current(self) -> Token:
        return self.tokens[self.pos]

    def peek(self, offset: int = 1) -> Optional[Token]:
        pos = self.pos + offset
        return self.tokens[pos] if pos < len(self.tokens) else None

    def advance(self) -> Token:
        token = self.tokens[self.pos]
        self.pos += 1
        return token

    def expect(self, token_type: TokenType) -> Token:
        token = self.current()
        if token.type != token_type:
            raise LMONParseError(
                f"Expected {token_type.value} but got {token.type.value}",
                token.line,
                token.column,
            )
        return self.advance()

    def skip_newlines(self) -> None:
        while self.current().type == TokenType.NEWLINE:
            self.advance()

    def parse(self) -> LMONDocument:
        self.skip_newlines()

        header: Optional[HeaderNode] = None

        if self.current().type == TokenType.LPAREN:
            header = self.parse_header()
            self.skip_newlines()

        body = self.parse_body()

        return LMONDocument(header=header, body=body)

    def parse_header(self) -> HeaderNode:
        self.expect(TokenType.LPAREN)
        labels = self.parse_label_list()
        self.expect(TokenType.RPAREN)
        return HeaderNode(labels=labels)

    def parse_label_list(self) -> List[LabelNode]:
        labels: List[LabelNode] = []

        labels.append(self.parse_label())

        while self.current().type == TokenType.COMMA:
            self.advance()
            labels.append(self.parse_label())

        return labels

    def parse_label(self) -> LabelNode:
        token = self.expect(TokenType.LABEL_OR_VALUE)
        is_array = False
        children: List[LabelNode] = []

        if self.current().type == TokenType.ARRAY_MARKER:
            self.advance()
            is_array = True

        # Check for nested schema: label:(sublabel,...)
        if self.current().type == TokenType.COLON:
            self.advance()
            if self.current().type == TokenType.LPAREN:
                self.advance()
                children = self.parse_label_list()
                self.expect(TokenType.RPAREN)

        return LabelNode(name=token.value, is_array=is_array, children=children)

    def parse_body(self) -> BodyNode:
        rows: List[RowNode] = []

        while self.current().type != TokenType.EOF:
            self.skip_newlines()
            if self.current().type == TokenType.EOF:
                break

            rows.append(self.parse_row())
            self.skip_newlines()

        return BodyNode(rows=rows)

    def parse_row(self) -> RowNode:
        label: Optional[str] = None

        # Check for row label (LABEL_OR_VALUE followed by COLON)
        next_token = self.peek()
        if (self.current().type == TokenType.LABEL_OR_VALUE and
                next_token and next_token.type == TokenType.COLON):
            label = self.advance().value
            self.expect(TokenType.COLON)

        # Allow bare array as row body, not just a document
        if self.current().type == TokenType.LBRACKET:
            array = self.parse_array()
            document = DocumentNode(fields=[array])
        else:
            document = self.parse_document()

        return RowNode(label=label, document=document)

    def parse_document(self) -> DocumentNode:
        self.expect(TokenType.LBRACE)

        fields: List[FieldValueNode] = []

        while self.current().type != TokenType.RBRACE:
            fields.append(self.parse_field_value())

            if self.current().type == TokenType.COMMA:
                self.advance()
                # Trailing comma is an error
                if self.current().type == TokenType.RBRACE:
                    raise LMONParseError(
                        "Trailing comma not allowed",
                        self.current().line,
                        self.current().column,
                    )

        self.expect(TokenType.RBRACE)

        return DocumentNode(fields=fields)

    def parse_field_value(self) -> FieldValueNode:
        if self.current().type == TokenType.LBRACE:
            return self.parse_document()

        if self.current().type == TokenType.LBRACKET:
            return self.parse_array()

        # Empty value case
        if (self.current().type == TokenType.COMMA or
                self.current().type == TokenType.RBRACE):
            return ValueNode(raw='')

        # Must be a value
        if (self.current().type == TokenType.LABEL_OR_VALUE or
                self.current().type == TokenType.QUOTED_STRING):
            token = self.advance()
            return ValueNode(raw=token.value)

        raise LMONParseError(
            f"Unexpected token: {self.current().type.value}",
            self.current().line,
            self.current().column,
        )

    def parse_array(self) -> ArrayNode:
        self.expect(TokenType.LBRACKET)

        items: List[FieldValueNode] = []

        while self.current().type != TokenType.RBRACKET:
            items.append(self.parse_field_value())

            if self.current().type == TokenType.COMMA:
                self.advance()
                if self.current().type == TokenType.RBRACKET:
                    raise LMONParseError(
                        "Trailing comma not allowed",
                        self.current().line,
                        self.current().column,
                    )

        self.expect(TokenType.RBRACKET)

        return ArrayNode(items=items)


def parse_to_ast(input_text: str) -> LMONDocument:
    """Parse LMON text to AST."""
    tokens = tokenize(input_text)
    parser = Parser(tokens)
    return parser.parse()


def parse(input_text: str) -> Union[Dict[str, Any], list]:
    """Parse LMON text to Python object."""
    ast = parse_to_ast(input_text)
    return evaluate(ast)


def evaluate(doc: LMONDocument) -> Union[Dict[str, Any], list]:
    """Evaluate AST to Python object."""
    if not doc.body.rows:
        return {} if doc.header else []

    # Check if rows have labels
    has_labels = any(row.label for row in doc.body.rows)
    all_have_labels = all(row.label for row in doc.body.rows)

    if has_labels and not all_have_labels:
        raise LMONParseError("Mixed labeled and unlabeled rows", 0, 0)

    if has_labels:
        result: Dict[str, Any] = {}
        for row in doc.body.rows:
            value = evaluate_document(row.document, doc.header.labels if doc.header else [])
            result[row.label] = value
        return result

    # No labels - return array
    result_array: list = []
    for row in doc.body.rows:
        value = evaluate_document(row.document, doc.header.labels if doc.header else [])
        result_array.append(value)
    return result_array


def evaluate_document(
    doc: DocumentNode,
    schema: List[LabelNode],
) -> Union[Dict[str, Any], list]:
    """Evaluate a document node."""
    values: list = []

    for field in doc.fields:
        schema_for_field = schema[len(values)].children if len(values) < len(schema) else []
        values.append(evaluate_field_value(field, schema_for_field))

    if not schema:
        # No schema: return array
        return values

    # Apply schema to values positionally
    result: Dict[str, Any] = {}
    for i, label in enumerate(schema):
        value = values[i] if i < len(values) else None

        if label.is_array:
            # This field should be an array
            if isinstance(value, list):
                result[label.name] = value
            elif value is None:
                result[label.name] = []
            else:
                result[label.name] = [value]
        elif label.children:
            # This field is a nested document
            result[label.name] = value
        else:
            result[label.name] = value

    return result


def evaluate_field_value(
    field: FieldValueNode,
    schema: List[LabelNode],
) -> Any:
    """Evaluate a field value."""
    if isinstance(field, ValueNode):
        return infer_type(field.raw)

    if isinstance(field, ArrayNode):
        return [evaluate_field_value(item, []) for item in field.items]

    if isinstance(field, DocumentNode):
        return evaluate_document(field, schema)

    return None
